File: go/0118_pascal_triangle.go
package main

func generate(numRows int) [][]int {
	if numRows == 0 {
		return [][]int{}
	} else if numRows == 1 {
		return [][]int{[]int{1}}
	}

	rows := [][]int{[]int{1}, []int{1, 1}}
	for i := 2; i < numRows; i++ {
		lastRow := rows[len(rows)-1]
		newRow := make([]int, len(lastRow)+1)
		for j := 0; j < len(newRow); j++ {
			if j == 0 || j == len(newRow)-1 {
				newRow[j] = 1
				continue
			}
			newRow[j] = lastRow[j] + lastRow[j-1]
		}
		rows = append(rows, newRow)
	}
	return rows
}

File: python/0118_pascal_triangle.py
class Solution:
    def generate(self, numRows):
        """
        :type numRows: int
        :rtype: List[List[int]]
        """
        if numRows == 0:
            return []

        res = [[1]]
        for i in range(1, numRows):
            prev = res[-1]
            if len(prev) == 1:
                res.append([1, 1])
                continue

            new = [1]
            j = 0
            while j + 1 < len(prev):
                new.append(prev[j] + prev[j + 1])
                j += 1
            new.append(1)
            res.append(new)
        return res